// github-action-ts-run-api/dist/runResult/AbstractRunResult.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractRunResult = void 0;
const FakeRunnerDir_1 = require("../githubServiceFiles/runnerDir/FakeRunnerDir");
class AbstractRunResult {
    constructor(commands, error, durationMs, isTimedOut, exitCode, stdout, stderr, runnerWarnings, tempDir, workspaceDir) {
        this.commands = commands;
        this.error = error;
        this.durationMs = durationMs;
        this.isTimedOut = isTimedOut;
        this.exitCode = exitCode;
        this.stdout = stdout;
        this.stderr = stderr;
        this.runnerWarnings = runnerWarnings;
        this.tempDir = tempDir;
        this.workspaceDir = workspaceDir;
        this.isSuccess = (this.exitCode === 0 || this.exitCode === undefined) && this.error === undefined;
    }
    get tempDirPath() {
        return this.tempDir.existingDirPath;
    }
    get workspaceDirPath() {
        return this.workspaceDir.existingDirPath;
    }
    cleanUpFakedDirs() {
        if (this.tempDir instanceof FakeRunnerDir_1.FakeRunnerDir) {
            this.tempDir.delete();
        }
        if (this.workspaceDir instanceof FakeRunnerDir_1.FakeRunnerDir) {
            this.workspaceDir.delete();
        }
        return this;
    }
}
exports.AbstractRunResult = AbstractRunResult;
