// github-action-ts-run-api/dist/runResult/warnings/WarningsCollector.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WarningsCollector = void 0;
const os_1 = __importDefault(require("os"));
class WarningsCollector {
    constructor(runOptions, actionConfig) {
        this.runOptions = runOptions;
        this.actionConfig = actionConfig;
        this.warnings = [];
        this.warnings.push(...this.actionConfig.getWarnings());
    }
    setCommandWarnings(warnings) {
        this.warnings.push(...warnings);
        return this;
    }
    /**
     * Get collected warnings and print if options.outputOptions.data.printRunnerWarnings is set
     */
    getAndPrint() {
        this.print();
        return this.get();
    }
    /**
     * Get collected warnings
     */
    get() {
        return this.warnings;
    }
    /**
     * Print if options.outputOptions.data.printRunnerWarnings is set
     */
    print() {
        if (this.runOptions.outputOptions.data.printRunnerWarnings) {
            this.warnings.forEach(warning => process.stderr.write(os_1.default.EOL + warning.message));
        }
    }
}
exports.WarningsCollector = WarningsCollector;
