// github-action-ts-run-api/dist/stdout/stdoutCommands.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.stdoutCmdRegexp = exports.stdoutCmdWithParamRegexp = exports.StdoutCommandName = void 0;
var StdoutCommandName;
(function (StdoutCommandName) {
    // noinspection SpellCheckingInspection,JSUnusedGlobalSymbols
    StdoutCommandName["ERROR"] = "error";
    StdoutCommandName["WARNING"] = "warning";
    StdoutCommandName["NOTICE"] = "notice";
    StdoutCommandName["DEBUG"] = "debug";
    StdoutCommandName["GROUP"] = "group";
    StdoutCommandName["END_GROUP"] = "endgroup";
    StdoutCommandName["SAVE_STATE"] = "save-state";
    StdoutCommandName["ADD_MASK"] = "add-mask";
    StdoutCommandName["ADD_PATH"] = "add-path";
    StdoutCommandName["ECHO"] = "echo";
    StdoutCommandName["SET_ENV"] = "set-env";
    StdoutCommandName["SET_OUTPUT"] = "set-output";
})(StdoutCommandName = exports.StdoutCommandName || (exports.StdoutCommandName = {}));
exports.stdoutCmdWithParamRegexp = /^::([A-Za-z\d\-_.]+?)(\s.+)?::(.*)?$/m;
exports.stdoutCmdRegexp = /^::([A-Za-z\d\-_.]+?)(\s.+)?::/m;
