// github-action-ts-run-api/dist/actionRunner/docker/runMilieu/DockerRunMilieu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DockerRunMilieu = void 0;
const BaseRunMilieu_1 = require("../../../runMilieu/BaseRunMilieu");
const RunnerDirsCollection_1 = require("../../../githubServiceFiles/RunnerDirsCollection");
class DockerRunMilieu {
    constructor(fakeFiles, runnerDirs, env, fakeFsOptions, volumes) {
        this.runnerDirs = runnerDirs;
        this.env = env;
        this.volumes = volumes;
        const baseRunnerDirs = new RunnerDirsCollection_1.RunnerDirsCollection({
            temp: runnerDirs.data.temp,
            workspace: runnerDirs.data.workspace
        });
        this._baseMilieu = new BaseRunMilieu_1.BaseRunMilieu(fakeFiles, baseRunnerDirs, env, fakeFsOptions);
    }
    getEffects() {
        return this._baseMilieu.getEffects(DockerRunMilieu.CONTAINER_OS_EOL);
    }
    restore() {
        this._baseMilieu.restore();
        this.runnerDirs.data.fileCommands.delete();
        this.runnerDirs.data.githubHome.delete();
        this.runnerDirs.data.githubWorkflow.delete();
    }
}
exports.DockerRunMilieu = DockerRunMilieu;
DockerRunMilieu.CONTAINER_OS_EOL = '\n';
