// github-action-ts-run-api/dist/runMilieu/BaseRunMilieu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseRunMilieu = void 0;
const FakeRunnerDir_1 = require("../githubServiceFiles/runnerDir/FakeRunnerDir");
class BaseRunMilieu {
    constructor(fakeFiles, runnerDirs, env, fakeFsOptions) {
        this.fakeFiles = fakeFiles;
        this.runnerDirs = runnerDirs;
        this.env = env;
        this.fakeFsOptions = fakeFsOptions;
    }
    getEffects(fileCommandsEol) {
        return {
            fileCommands: this.fakeFiles.readFileCommands(fileCommandsEol),
            runnerDirs: this.runnerDirs
        };
    }
    restore() {
        this.cleanUpFakeFiles();
        this.cleanUpFakeDirs();
    }
    cleanUpFakeFiles() {
        this.fakeFiles.cleanUp();
    }
    cleanUpFakeDirs() {
        if (this.fakeFsOptions.rmFakedTempDirAfterRun && this.runnerDirs.data.temp instanceof FakeRunnerDir_1.FakeRunnerDir) {
            this.runnerDirs.data.temp.delete();
        }
        if (this.fakeFsOptions.rmFakedWorkspaceDirAfterRun && this.runnerDirs.data.workspace instanceof FakeRunnerDir_1.FakeRunnerDir) {
            this.runnerDirs.data.workspace.delete();
        }
    }
}
exports.BaseRunMilieu = BaseRunMilieu;
