// github-action-ts-run-api/dist/actionRunner/fn/runMilieu/FnRunMilieu.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FnRunMilieu = void 0;
const BaseRunMilieu_1 = require("../../../runMilieu/BaseRunMilieu");
const os_1 = __importDefault(require("os"));
class FnRunMilieu {
    constructor(fakeFiles, runnerDirs, env, fakeFsOptions, restoreProcessProps, stdoutInterceptor) {
        this.env = env;
        this.restoreProcessProps = restoreProcessProps;
        this.stdoutInterceptor = stdoutInterceptor;
        this._baseRunMilieu = new BaseRunMilieu_1.BaseRunMilieu(fakeFiles, runnerDirs, env, fakeFsOptions);
    }
    getEffects() {
        const baseEffects = this._baseRunMilieu.getEffects(os_1.default.EOL);
        this.stdoutInterceptor.finishCommandsParsing();
        return Object.assign(Object.assign({}, baseEffects), { stdoutCommands: this.stdoutInterceptor.parsedCommands.data, exitCode: process.exitCode, stdout: this.stdoutInterceptor.interceptedStdout, stderr: this.stdoutInterceptor.interceptedStderr });
    }
    restore() {
        this.stdoutInterceptor.unHook();
        this.restoreProcessProps();
        this._baseRunMilieu.restore();
    }
}
exports.FnRunMilieu = FnRunMilieu;
